### PR TITLE
Add SingleKey and IfEmptyObject

### LINKED ISSUE #813: Require exactly one key of type `string`
Is there was a way to type an object to require that it has exactly one key? The key itself can be any `string` (or `number` or `symbol` for that matter), but I'd like to ensure there are at least _and_ at most one non-prototype/inherited property.

The use case is for typing the JavaScript implementation of [JsonLogic](http://jsonlogic.com/) ([npm](https://www.npmjs.com/package/json-logic-js) / [DT types](https://www.npmjs.com/package/@types/json-logic-js)). In the runtime code, the "logic" object itself and each nested object must satisfy the condition `Object.keys(logic).length === 1` ([see code](https://github.com/jwadhams/json-logic-js/blob/2.0.2/logic.js#L180)).

I tried `RequireExactlyOne<Record<string, any>, string>`, but that seems to be effectively the same as `Record<string, any>`.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/813">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/813/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by ash-zzz**
Closes #813 

I also added IfEmptyObject since that was a missing guard type that I needed to use for this

### CODE REVIEW

**Comment by sindresorhus** [comment]
I think it should be named `SingleKeyObject`. The current name is a bit ambiguous and could make you single it fetches a single key from an object, for example.

File: source/single-key.d.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+processOperation({ operation: { name: 'add' } });
+
+processOperation({ operation: { name: 'add' }, 'values': [1, 2] }); // Compliation error
+```
+
+@category Object
+ */
+
+// export type SingleKey<ObjectType> =
+// 	IsUnion<keyof ObjectType> extends true ? never : {} extends ObjectType ? never : ObjectType;
```
Comment (sindresorhus) [inline source/single-key.d.ts:32]: Leftover

File: source/single-key.d.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+const processOperation = <T>(operation: SingleKey<T>) => {};
+
+processOperation({ operation: { name: 'add' } });
+
+processOperation({ operation: { name: 'add' }, 'values': [1, 2] }); // Compliation error
+```
+
+@category Object
+ */
```
Comment (sindresorhus) [inline source/single-key.d.ts:29]:
```suggestion
*/
```

File: source/single-key.d.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	}
+});
+
+processOperation({
+	operation: {
+		name: 'add'
+	},
+	values: [1, 2]
+});
+// Compilation error
```
Comment (sindresorhus) [inline source/single-key.d.ts:25]: Document the actual compilation here.

File: source/single-key.d.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+processOperation({
+	operation: {
+		name: 'add'
+	}
+});
+
+processOperation({
+	operation: {
+		name: 'add'
+	},
```
Comment (sindresorhus) [inline source/single-key.d.ts:22]: Keep the example simple by not using a nested property here.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/single-key.d.ts (hunk 1 above)
Comment (ash-zzz) [reply source/single-key.d.ts]: Oops I thought I removed them all

**Comment by ash-zzz** [comment]
> I think it should be named `SingleKeyObject`. The current name is a bit ambiguous and could make you single it fetches a single key from an object, for example.

Agreed, I just renamed it and fixed the documentation

**Comment by sindresorhus** [comment]
Thanks :)

**Comment by jakeboone02** [comment]
Thanks for doing this, @ash-zzz!

Should this work for types where the keys are unrestricted? If not, I'm not sure it satisfies my original request from #813, which was to declare an object type that requires exactly one key without restricting the key itself to any specific string or union of strings.

I tried `SingleKeyObject<Record<string, any>>` and `SingleKeyObject<{ [k: string]: any }>` but they both resolved to `never`. Something like `SingleKeyObject<Record<'union' | 'of' | 'strings', any>>` works fine.

**Pull request merged**